File: src/control-tools.ts
import { createHash } from "node:crypto"
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js"
import { keyValueSetArgs, namespaceArgs } from "./args.js"
import {
  HelmUpgradeConfirmInput,
  HelmUpgradePlanInput,
  KubectlApplyConfirmInput,
  KubectlApplyPlanInput
} from "./schemas.js"
import { commandTool } from "./tool-runner.js"
import type { ChangePlan, PlanKind, RuntimeConfig } from "./types.js"
import { respond } from "./response.js"

type KubectlApplyCommand = {
  readonly cluster: string
  readonly namespace: string | undefined
  readonly manifestPath: string
  readonly dryRun: boolean
}

type HelmCommand = {
  readonly cluster: string
  readonly namespace: string | undefined
  readonly release: string
  readonly chart: string
  readonly valuesFile: string | undefined
  readonly set: Readonly<Record<string, string | number | boolean>>
}

type HelmUpgradeCommand = HelmCommand & {
  readonly install: boolean
}

type PlanSeed = {
  readonly kind: PlanKind
  readonly command: string
  readonly args: readonly string[]
  readonly warnings: readonly string[]
}

type GateCheck = {
  readonly config: RuntimeConfig
  readonly plan: ChangePlan
  readonly confirmationHash: string
  readonly confirmationToken: string
}

export function registerControlTools(server: McpServer, config: RuntimeConfig): void {
  server.registerTool(
    "plan_kubectl_apply",
    { description: "Plan kubectl apply with local client-side dry-run only.", inputSchema: KubectlApplyPlanInput.shape },
    ({ cluster, namespace, manifestPath }) =>
      commandTool(config, { command: "kubectl", args: kubectlApplyArgs({ cluster, namespace, manifestPath, dryRun: true }) }).then((response) => {
        const plan = createPlan({
          kind: "kubectl_apply",
          command: "kubectl",
          args: kubectlApplyArgs({ cluster, namespace, manifestPath, dryRun: false }),
          warnings: [
            "This plan uses --dry-run=client and does not contact the cluster by default.",
            "A human may run a separate server-side dry-run only after selecting the intended cluster context.",
            "Apply tools only run when EKS_OPS_MODE=control."
          ]
        })
        return respond({ ...response.structuredContent, plan })
      })
  )

  server.registerTool(
    "apply_kubectl_apply_confirmed",
    { description: "Apply a manifest only in control mode with a matching confirmation hash/token.", inputSchema: KubectlApplyConfirmInput.shape },
    ({ cluster, namespace, manifestPath, confirmationHash, confirmationToken }) => {
      const args = kubectlApplyArgs({ cluster, namespace, manifestPath, dryRun: false })
      const plan = createPlan({ kind: "kubectl_apply", command: "kubectl", args, warnings: [] })
      const gate = checkGate({ config, plan, confirmationHash, confirmationToken })
      return gate === undefined ? commandTool(config, { command: "kubectl", args, intent: "confirmed_control" }) : Promise.resolve(gate)
    }
  )

  server.registerTool(
    "plan_helm_upgrade",
    { description: "Plan Helm upgrade by rendering templates locally.", inputSchema: HelmUpgradePlanInput.shape },
    ({ cluster, namespace, release, chart, valuesFile, set, install }) =>
      commandTool(config, { command: "helm", args: helmTemplateArgs({ cluster, namespace, release, chart, valuesFile, set }) }).then((response) => {
        const plan = createPlan({
          kind: "helm_upgrade",
          command: "helm",
          args: helmUpgradeArgs({ cluster, namespace, release, chart, valuesFile, set, install }),
          warnings: [
            "Planning uses helm template, which renders locally and does not install or upgrade the release.",
            "Review rendered manifests and helm diff externally if required by policy.",
            "Apply tools only run when EKS_OPS_MODE=control."
          ]
        })
        return respond({ ...response.structuredContent, plan })
      })
  )

  server.registerTool(
    "apply_helm_upgrade_confirmed",
    { description: "Run helm upgrade only in control mode with a matching confirmation hash/token.", inputSchema: HelmUpgradeConfirmInput.shape },
    ({ cluster, namespace, release, chart, valuesFile, set, install, confirmationHash, confirmationToken }) => {
      const args = helmUpgradeArgs({ cluster, namespace, release, chart, valuesFile, set, install })
      const plan = createPlan({ kind: "helm_upgrade", command: "helm", args, warnings: [] })
      const gate = checkGate({ config, plan, confirmationHash, confirmationToken })
      return gate === undefined ? commandTool(config, { command: "helm", args, intent: "confirmed_control" }) : Promise.resolve(gate)
    }
  )
}

function kubectlApplyArgs(command: KubectlApplyCommand): readonly string[] {
  return [
    "--context",
    command.cluster,
    "apply",
    "--filename",
    command.manifestPath,
    ...namespaceArgs(command.namespace),
    ...(command.dryRun ? ["--dry-run=client", "--output", "yaml"] : [])
  ]
}

function helmTemplateArgs(command: HelmCommand): readonly string[] {
  return [
    "template",
    command.release,
    command.chart,
    "--kube-context",
    command.cluster,
    ...namespaceArgs(command.namespace),
    ...(command.valuesFile === undefined ? [] : ["--values", command.valuesFile]),
    ...keyValueSetArgs(command.set)
  ]
}

function helmUpgradeArgs(command: HelmUpgradeCommand): readonly string[] {
  return [
    "upgrade",
    command.release,
    command.chart,
    "--kube-context",
    command.cluster,
    ...namespaceArgs(command.namespace),
    ...(command.install ? ["--install"] : []),
    ...(command.valuesFile === undefined ? [] : ["--values", command.valuesFile]),
    ...keyValueSetArgs(command.set)
  ]
}

function createPlan(seed: PlanSeed): ChangePlan {
  const confirmationHash = createHash("sha256").update(JSON.stringify({ command: seed.command, args: seed.args })).digest("hex")
  return { ...seed, confirmationHash, confirmationToken: `confirm:${confirmationHash}` }
}

function checkGate(check: GateCheck): ReturnType<typeof respond> | undefined {
  if (check.config.mode !== "control") {
    return respond({ status: "blocked", reason: "Apply tools are disabled unless EKS_OPS_MODE=control.", plan: check.plan })
  }
  if (check.confirmationHash !== check.plan.confirmationHash || check.confirmationToken !== check.plan.confirmationToken) {
    return respond({ status: "blocked", reason: "Confirmation hash/token did not match the planned command.", plan: check.plan })
  }
  return undefined
}

File: safe_eks_mcp/control_tools.py
from __future__ import annotations

import hashlib
import json

from .args import key_value_set_args, namespace_args
from .helm_tools import helm_template
from .response import change_plan_dict
from .runner import command_tool
from .types import ChangePlan, ChangePlanJson, CliCommand, CliInvocation, PlanKind, RuntimeConfig, ToolPayloadJson
from .validation import (
    normalize_set_values,
    optional_namespace,
    require_cluster,
    require_name,
    require_path,
    require_region,
)


def plan_kubectl_apply(
    config: RuntimeConfig, region: str, cluster: str, manifestPath: str, namespace: str | None = None
) -> ToolPayloadJson:
    _ = require_region(region)
    dry_run_args = kubectl_apply_args(cluster, namespace, manifestPath, dry_run=True)
    response = command_tool(config, CliInvocation(command="kubectl", args=dry_run_args))
    plan = create_plan(
        "kubectl_apply",
        "kubectl",
        kubectl_apply_args(cluster, namespace, manifestPath, dry_run=False),
        (
            "This plan uses --dry-run=client and does not contact the cluster by default.",
            "A human may run a separate server-side dry-run only after selecting the intended cluster context.",
            "Apply tools only run when EKS_OPS_MODE=control.",
        ),
    )
    response["plan"] = payload_dict_plan(plan)
    return response


def apply_kubectl_apply_confirmed(
    config: RuntimeConfig,
    region: str,
    cluster: str,
    manifestPath: str,
    confirmationHash: str,
    confirmationToken: str,
    namespace: str | None = None,
) -> ToolPayloadJson:
    _ = require_region(region)
    args = kubectl_apply_args(cluster, namespace, manifestPath, dry_run=False)
    plan = create_plan("kubectl_apply", "kubectl", args, ())
    gate = check_gate(config, plan, confirmationHash, confirmationToken)
    if gate is not None:
        return gate
    return command_tool(config, CliInvocation(command="kubectl", args=args, intent="confirmed_control"))


def plan_helm_upgrade(
    config: RuntimeConfig,
    region: str,
    cluster: str,
    release: str,
    chart: str,
    namespace: str | None = None,
    valuesFile: str | None = None,
    set: dict[str, str | int | bool] | None = None,
    install: bool = True,
) -> ToolPayloadJson:
    response = helm_template(config, region, cluster, release, chart, namespace, valuesFile, set)
    plan = create_plan(
        "helm_upgrade",
        "helm",
        helm_upgrade_args(cluster, namespace, release, chart, valuesFile, set, install),
        (
            "Planning uses helm template, which renders locally and does not install or upgrade the release.",
            "Review rendered manifests and helm diff externally if required by policy.",
            "Apply tools only run when EKS_OPS_MODE=control.",
        ),
    )
    response["plan"] = payload_dict_plan(plan)
    return response


def apply_helm_upgrade_confirmed(
    config: RuntimeConfig,
    region: str,
    cluster: str,
    release: str,
    chart: str,
    confirmationHash: str,
    confirmationToken: str,
    namespace: str | None = None,
    valuesFile: str | None = None,
    set: dict[str, str | int | bool] | None = None,
    install: bool = True,
) -> ToolPayloadJson:
    _ = require_region(region)
    args = helm_upgrade_args(cluster, namespace, release, chart, valuesFile, set, install)
    plan = create_plan("helm_upgrade", "helm", args, ())
    gate = check_gate(config, plan, confirmationHash, confirmationToken)
    if gate is not None:
        return gate
    return command_tool(config, CliInvocation(command="helm", args=args, intent="confirmed_control"))


def kubectl_apply_args(cluster: str, namespace: str | None, manifest_path: str, dry_run: bool) -> tuple[str, ...]:
    args = [
        "--context",
        require_cluster(cluster),
        "apply",
        "--filename",
        require_path(manifest_path, "manifestPath"),
        *namespace_args(optional_namespace(namespace)),
    ]
    if dry_run:
        args.extend(("--dry-run=client", "--output", "yaml"))
    return tuple(args)


def helm_upgrade_args(
    cluster: str,
    namespace: str | None,
    release: str,
    chart: str,
    values_file: str | None,
    set_values: dict[str, str | int | bool] | None,
    install: bool,
) -> tuple[str, ...]:
    args = [
        "upgrade",
        require_name(release, "release"),
        require_path(chart, "chart"),
        "--kube-context",
        require_cluster(cluster),
        *namespace_args(optional_namespace(namespace)),
    ]
    if install:
        args.append("--install")
    if values_file is not None:
        args.extend(("--values", require_path(values_file, "valuesFile")))
    args.extend(key_value_set_args(normalize_set_values(set_values)))
    return tuple(args)


def create_plan(kind: PlanKind, command: CliCommand, args: tuple[str, ...], warnings: tuple[str, ...]) -> ChangePlan:
    encoded = json.dumps({"command": command, "args": list(args)}, separators=(",", ":"))
    confirmation_hash = hashlib.sha256(encoded.encode("utf-8")).hexdigest()
    return ChangePlan(
        kind=kind,
        command=command,
        args=args,
        confirmation_hash=confirmation_hash,
        confirmation_token=f"confirm:{confirmation_hash}",
        warnings=warnings,
    )


def check_gate(
    config: RuntimeConfig, plan: ChangePlan, confirmation_hash: str, confirmation_token: str
) -> ToolPayloadJson | None:
    if config.mode != "control":
        return {
            "status": "blocked",
            "reason": "Apply tools are disabled unless EKS_OPS_MODE=control.",
            "plan": payload_dict_plan(plan),
        }
    if confirmation_hash != plan.confirmation_hash or confirmation_token != plan.confirmation_token:
        return {
            "status": "blocked",
            "reason": "Confirmation hash/token did not match the planned command.",
            "plan": payload_dict_plan(plan),
        }
    return None


def payload_dict_plan(plan: ChangePlan) -> ChangePlanJson:
    return change_plan_dict(plan)
